Allow other depths

zhaori96 opened the pull request:
### Description

This pull request addresses [Issue #83 ](https://github.com/boombuler/barcode/issues/83), which proposes an enhancement to support generating barcodes with a specified image depth. The changes introduced in this PR allow users to specify the image depth when generating barcodes, providing greater flexibility and performance improvements for various applications.

### Changes Made

- Added utility functions in the `utils` package to handle different color models and color depths:
  - `ColorModel(depth int) color.Model`
  - `WhiteColor(depth int) color.Color`
  - `BlackColor(depth int) color.Color`
- Updated the `base1DCode` struct to use these utility functions for color model and color selection based on the specified depth.
- Modified all supported barcode types to include an `EncodeWithDepth` function that allows specifying the image depth during barcode generation.

### Example Usage

```go
barcode, err := somebarcodepackage.EncodeWithDepth("123456", true, 32)
if err != nil {
    // handle error
}
// Use the barcode...
```

### Benefits

- **Performance Improvement**: Reduces the overhead associated with image conversion by generating images with the required depth directly.
- **Flexibility**: Allows users to generate barcodes that better fit their specific use cases and integration needs.
- **Simplicity**: Minimizes the need for additional image processing code.

### Reference

This PR resolves [Issue #83](https://github.com/boombuler/barcode/issues/83).

Thank you for considering this enhancement. I believe it will make the boombuler/barcode package more versatile and efficient for a wider range of applications.

boombuler commented:
Do you have benchmarks how much faster this is then converting the image "by hand"?
Because changing the color model is almost just 2 lines of code, and is imho even more flexible.

zhaori96 commented:
**Benchmark Results:**

```
BenchmarkAutoDepth-8              121581              9318 ns/op
BenchmarkManualDepth-8              2005            571702 ns/op
```

**Benchmark Code:**

```go
package main

import (
	"image"
	"image/draw"
	"testing"

	"github.com/boombuler/barcode"
	"github.com/boombuler/barcode/twooffive"
)

// BenchmarkAutoDepth measures the execution time of AutoDepth
func BenchmarkAutoDepth(b *testing.B) {
	for i := 0; i < b.N; i++ {
		img, _ := twooffive.EncodeWithDepth("341900900000002600849320205700011000", true, 32)
		img, _ = barcode.Scale(img, 400, 50)
	}
}

// BenchmarkManualDepth measures the execution time of ManualDepth
func BenchmarkManualDepth(b *testing.B) {
	for i := 0; i < b.N; i++ {
		img, _ := twooffive.Encode("341900900000002600849320205700011000", true)
		img, _ = barcode.Scale(img, 400, 50)

		bounds := img.Bounds()
		dst := image.NewRGBA(bounds)
		draw.Draw(dst, bounds, img, bounds.Min, draw.Src)
	}
}
```

zhaori96 commented:
Hi,

Is there anything else I should do to improve the PR or any additional information needed? I am keen to understand if there is a possibility of the PR being accepted.

Thank you for your time and consideration.

boombuler commented:
Hey sorry for the late response but I only have my phone to view this for now. But what would you think about passing in a color model and two colors instead? That way you would have even more options and could easily create colored barcodes for example

zhaori96 commented:
> Hey sorry for the late response but I only have my phone to view this for now. But what would you think about passing in a color model and two colors instead? That way you would have even more options and could easily create colored barcodes for example

I’ve implemented the changes you suggested and aimed to maintain flexibility in adjusting the pixel density of the generated barcode. To achieve this, I introduced the ColorScheme struct along with default instances for 8-bit, 16-bit, 24-bit, and 32-bit color schemes.

Additionally, I noticed that the Scale function fills the excess area of the resized image with a fixed white color. To address this, I added a new function, ScaleWithFill, which allows specifying a fill color other than white, or even setting it to transparent if needed.

Please let me know if there are any further improvements or adjustments required.

boombuler commented on code128/encode.go:
```diff
@@ -180,7 +180,12 @@ func Encode(content string) (barcode.BarcodeIntCS, error) {
+	return EncodeWithColor(content, barcode.ColorScheme16)
 }
 
 func EncodeWithoutChecksum(content string) (barcode.Barcode, error) {
```
This Encoder is missing a color overload

zhaori96 replied:
I didn't understand what you meant. Apparently this encoder is the same as the others. Encode calls EncodeWithColor passing barcode. ColorScheme16 in a fixed way. I did not change the expected parameters of the Encode method to maintain compatibility.

boombuler replied:
I was talking about the `EncodeWithoutChecksum` overload. Imho there should also be a variant to encode with color and without checksum

zhaori96 replied:
Overload created in [commit a8e67c5](https://github.com/boombuler/barcode/pull/84/commits/a8e67c5d16f5f4de67a62c2dd28d48f9e52216c6)

boombuler commented on scaledbarcode.go:
```diff
@@ -49,11 +49,16 @@ func (bc *intCSscaledBC) CheckSum() int {
 
 // Scale returns a resized barcode with the given width and height.
 func Scale(bc Barcode, width, height int) (Barcode, error) {
+	return ScaleWithFill(bc, width, height, color.White)
```
I would suggest using the background of the used colorscheme instead of hardcoded white.

zhaori96 replied:
To achieve this, I would need to request the ColorScheme via a parameter in the Scale function, which would be problematic due to the incompatibilities it generates.

I considered modifying the Barcode interface to add the ColorScheme() barcode.ColorScheme method, but I am unsure if it is appropriate to change this interface.

boombuler replied:
You could add another interface to access the background. That way the original interface does not need to be changed. If the barcode is not implementing the new interface, you could then fallback to white

boombuler commented on utils/depth.go:
```diff
@@ -0,0 +1,36 @@
+
+import "image/color"
+
+func ColorModel(depth int) color.Model {
```
Is this file still used?

zhaori96 replied:
I forgot to remove it. I will do this.

zhaori96 replied:
File removed in [commit 1c82249](https://github.com/zhaori96/barcode/commit/1c8224977e1efd6f585a2fb6a20809b6af6b5dd8)

boombuler reviewed (commented):
The rest looks  good to me. Thank you for contributing 👍
